kernel_dmesg_diff: Add command-line option for previous kernel dmesg logs

pratheekshasn opened the pull request:
### Summary of Changes

Add the ability to set a particular log as the basis against which the current build's kernel_dmesg_diff test log should be compared against. This is so that newer kernel dmesg outputs can be accepted without triggering the test failure for innocuous changes in the kernel dmesg output.


### Justification

[#AB2990690](https://dev.azure.com/ni/DevCentral/_workitems/edit/2990690/)

### Testing

* [x] I have built the core package feed with this PR in place. (`bitbake packagefeed-ni-core`)
* [x] Built the package and installed it on a target, ran the test and verified that the test can fetch records from the database based on the date provided. It also generates the logs.

### Procedure
* [x] I certify that the contents of this pull request complies with the [Developer Certificate of Origin](https://github.com/ni/nilrt/blob/HEAD/docs/CONTRIBUTING.md#developer-certificate-of-origin-dco).

chaitu236 reviewed (changes requested):
Changes look fine. But please fix the commit message to follow 50/72 rule.

Also mention the testing you did to ensure these changes work (and not just the generic `bitbake packagefeed-ni-core` build).

jpautler commented on recipes-kernel/kernel-tests/kernel-tests-files/kernel_dmesg_diff.py:
```diff
@@ -284,6 +284,10 @@ def parse_args():
     parser.add_argument('--skip_upload', help='Skip upload of dmesg record to database. Useful when debugging.', action="store_true")
     parser.add_argument('--suppress_diff_output', help='Reduces test output noise by removing the dmesg log diff from the output. Useful when debugging.', action="store_true")
+    parser.add_argument('--basis_log_db_date', metavar="<date>",
+                        help='Use this flag to supply a date string that will be used to locate a kernel dmesg log in the database. That log will be used as the basis kernel dmseg log. '\
```
Typo: basis kernel **dmseg** log

pratheekshasn replied:
I have fixed this now.

jpautler commented on recipes-kernel/kernel-tests/kernel-tests-files/kernel_dmesg_diff.py:
```diff
@@ -304,6 +308,10 @@ def parse_args():
     previous_dmesg_record = get_previous_dmesg_record(db, kernel_version, os_version.major_minor, device_desc, logger)
 
+if args.basis_log_db_date:
+    previous_dmesg_record = get_dmesg_record_by_date(db, args.basis_log_db_date, logger)
```
Rather than set previous_dmesg_record above (in the if/else cases of args.current_log_db_date) and then just overwrite it here, you should add an else case here to set previous_dmesg_record=get_dmesg_record_by_date(...). This structure will make this if/else mimic the one above, just for the basis log instead of current log.

pratheekshasn replied:
I have cleaned this up now.

jpautler commented on recipes-kernel/kernel-tests/kernel-tests-files/kernel_dmesg_diff.py:
```diff
@@ -295,13 +299,17 @@ def parse_args():
+    previous_dmesg_record = get_dmesg_record_by_date(db, args.basis_kernel_dmesg_log_db_date, logger)
+    assert previous_dmesg_record, "Could not find matching basis log record from database."
+else:
     previous_dmesg_record = get_previous_dmesg_record(db, kernel_version, os_version.major_minor, device_desc, logger)
```
This code no longer correctly locates a previous dmesg record when the caller specifies only --current_log_db_date.

When testing, you should be able to call this script from your desktop machine and pass in a current_log_db_date from a prior test failure and ensure you can replicate the same output as the prior test failure.

pratheekshasn replied:
It does locate a previous dmesg record, right? If only current_log_db_date is specified, it gets the previous_dmesg_record in the else case of basis_kernel_dmesg_log_db_date. This else case runs regardless of specifying current_log_db_date.

pratheekshasn replied:
I have also run the test by specifying only the current_log_db_date, and it does call the get_previous_dmesg_record() function.

jpautler replied:
I can see that you are calling get_previous_dmesg_record in this case, but you are no longer **correctly** finding the previous dmesg record. You are finding some other incorrect previous dmesg record. Please run the script both before and after your change with **--current_log_db_date "2025-01-15 05:43:51.687559"** and note that the previous ("basis") records being found are not the same.

pratheekshasn replied:
Here is the beginning of the log with my changes:

```
INFO: dmesg_diff: 6.6.63-rt46 against <empty>, diff hash 711cc18e692baa7e0691bf67b369538f 

INFO: No suitable previous log found. The following query was used: 

INFO: {'kernel_version': {'$lt': {'major': 6, 'minor': 6, 'patch': 63, 'rc': 999, 'rt_patch': 46}}, 'device_desc': 'cRIO-9045'} 

INFO: Uploaded dmesg log of "next" kernel 6.6.63-rt46-next-gbf84a0bee986 from 2025-03-03 07:00:50.514829 with _id 67c553a2fe0153d062f63663 

INFO: Starting diff
```

Here is the log without my changes:

```
INFO: dmesg_diff: 6.6.63-rt46 against <empty>, diff hash 711cc18e692baa7e0691bf67b369538f 

INFO: No suitable previous log found. The following query was used: 

INFO: {'kernel_version': {'$lt': {'major': 6, 'minor': 6, 'patch': 63, 'rc': 999, 'rt_patch': 46}}, 'device_desc': 'cRIO-9045'} 

INFO: Uploaded dmesg log of "next" kernel 6.6.63-rt46-next-gbf84a0bee986 from 2025-03-03 07:00:50.514829 with _id 67c553a2fe0153d062f63663 

INFO: Starting diff
```

pratheekshasn replied:
However, I have now changed the code, and this is the log:

```
INFO: dmesg_diff: 6.6.63-rt46 against 6.6.52-rt43, diff 099eabfa0d07ab9fa01ffa02e43af3d6 

INFO: Found dmesg record with date: 2025-01-15 05:43:51.687559 with kernel version: 6.6.63-rt46-next-gbf84a0bee986 and OS version: 25.0.0f163-x64 

INFO: Most recent previous kernel version in database is 6.6.52-rt43. 

INFO: Using previous dmesg log of "next" kernel 6.6.52-rt43-next-g3254def04a55 from OS version 25.0.0d135-x64 dated 2024-12-10 18:27:50.505027 with _id 67588826f8342ecde14c2d68 

INFO: Starting diff
```

jpautler commented on recipes-kernel/kernel-tests/kernel-tests-files/kernel_dmesg_diff.py:
```diff
@@ -284,6 +284,10 @@ def parse_args():
                              'database, so the --skip_upload flag is not needed. This flag is useful for debugging issues with this test.')
     parser.add_argument('--skip_upload', help='Skip upload of dmesg record to database. Useful when debugging.', action="store_true")
     parser.add_argument('--suppress_diff_output', help='Reduces test output noise by removing the dmesg log diff from the output. Useful when debugging.', action="store_true")
+    parser.add_argument('--basis_kernel_dmesg_log_db_date', metavar="<date>",
```
Change this argument name from basis_kernel_dmesg_log_db_date to basis_log_db_date. This better matches current_log_db_date. (The only difference between these option names should be to replace "current" with "basis".)

pratheekshasn replied:
Fixed.

jpautler commented on recipes-kernel/kernel-tests/kernel-tests-files/kernel_dmesg_diff.py:
```diff
@@ -284,25 +284,37 @@ def parse_args():
-
     kernel_version = KernelVersion()
-    os_version = OsVersion()
+    os_version = OsVersion().major_minor
```
You have introduced another bug here by changing the type of value stored in os_version to only be the major_minor field. The os_version variable is used again below in the call to upload_log, so you cannot change the type of the value.

pratheekshasn replied:
Sorry about that Jeff, I was playing around with something, and forgot to remove this. I have fixed this now.

pratheekshasn replied:
Here are the logs I got from running the test on a target:

With current_date specified:
```
INFO: dmesg_diff: 6.6.63-rt46 against 6.6.52-rt43, diff 099eabfa0d07ab9fa01ffa02e43af3d6 

INFO: Found dmesg record with date: 2025-01-15 05:43:51.687559 with kernel version: 6.6.63-rt46-next-gbf84a0bee986 and OS version: 25.0.0f163-x64 

INFO: Most recent previous kernel version in database is 6.6.52-rt43. 

INFO: Using previous dmesg log of "next" kernel 6.6.52-rt43-next-g3254def04a55 from OS version 25.0.0d135-x64 dated 2024-12-10 18:27:50.505027 with _id 67588826f8342ecde14c2d68 

INFO: Starting diff
```
With basis_date specified: 
```
INFO: Found dmesg record with date: 2025-01-15 05:43:51.687559 with kernel version: 6.6.63-rt46-next-gbf84a0bee986 and OS version: 25.0.0f163-x643d851e25204c4551b568ae7ed9f6958e 

INFO: Uploaded dmesg log of "next" kernel 6.6.63-rt46-next-gbf84a0bee986 from 2025-03-04 07:21:38.326773 with _id 67c6aa02a8947b70b2cfad83 

INFO: Starting diff
```
With both: 
```
INFO: Found dmesg record with date: 2025-01-15 05:43:51.687559 with kernel version: 6.6.63-rt46-next-gbf84a0bee986 and OS version: 25.0.0f163-x6402075dcb8ddc114d953537552f2ba44e 

INFO: Found dmesg record with date: 2025-01-15 08:49:03.205847 with kernel version: 6.6.63-rt46-next-gbf84a0bee986 and OS version: 25.0.0f163-x64 

INFO: Starting diff
```
With neither: 
```
INFO: dmesg_diff: 6.6.63-rt46 against <empty>, diff hash 711cc18e692baa7e0691bf67b369538f 

INFO: No suitable previous log found. The following query was used: 

INFO: {'kernel_version': {'$lt': {'major': 6, 'minor': 6, 'patch': 63, 'rc': 999, 'rt_patch': 46}}, 'device_desc': 'cRIO-9045'} 

INFO: Uploaded dmesg log of "next" kernel 6.6.63-rt46-next-gbf84a0bee986 from 2025-03-04 08:02:30.395285 with _id 67c6b396549a89c2a66aa664 

INFO: Starting diff
```

jpautler commented on recipes-kernel/kernel-tests/kernel-tests-files/kernel_dmesg_diff.py:
```diff
@@ -284,25 +284,37 @@ def parse_args():
     parser.add_argument('--skip_upload', help='Skip upload of dmesg record to database. Useful when debugging.', action="store_true")
     parser.add_argument('--suppress_diff_output', help='Reduces test output noise by removing the dmesg log diff from the output. Useful when debugging.', action="store_true")
+    parser.add_argument('--basis_log_db_date', metavar="<date>",
+                        help='Use this flag to supply a date string that will be used to locate a kernel dmesg log in the database. That log will be used as the basis kernel dmseg log. '\
```
I would remove all occurrences of "kernel" from this description. It is already implied that the dmesg log contains output from the kernel.

Otherwise, everything looks good.

pratheekshasn replied:
Fixed this now.

CommitSentience committed:
kernel_dmesg_diff: Add command-line option for previous kernel dmesg logs

Add the ability to specify a baseline log against which the current
build's kernel_dmesg_diff test log is compared. This allows newer
kernel dmesg outputs to be accepted without triggering test failures
due to innocuous changes in the kernel dmesg output.

This helps reduce noise in test failures caused by expected or minor
changes in kernel messages while still catching unexpected regressions.

Signed-off-by: Pratheeksha S N <pratheeksha.s.n@emerson.com>
